hystrix/hystrix.go

```diff
@@ -20,6 +20,29 @@ func (e CircuitError) Error() string {
 	return "hystrix: " + e.Message
 }
 
+// A BadRequest is an error which is treated as a successful call.
+// The fallback function will not be executed when a BadRequest is returned.
+// BadRequests do not count against failure metrics
+type BadRequest interface {
```

Comment:
This can be another error like `ErrCircuitOpen`, `ErrMaxConcurrency`etc.
ref: https://github.com/myteksi/hystrix-go/blob/master/hystrix/hystrix.go#L46

Reply:
Not sure if I follow here.

None of the `CircuitErrors` will match this interface since they do not implement the `BadRequest` method.

Would you like to see some tests to assert that those errors cannot be asserted as BadRequests?

Reply:
Are you looking for another `CircuitError` variable? I'm not sure that that is in line with @epivovar original request [here](https://github.com/afex/hystrix-go/issues/72) as an implementation based on a struct would not allow for the various errors implemented in client packages to be treated as `BadRequests` themselves

Reply:
My bad, all this time I though https://godoc.org/github.com/pkg/errors is part of the official errors package.
Here, is my opinion:

Create ErrBadRequest is a type of CircuitError `ErrBadRequest = CircuitError{Message: "BadRequest"}`
Then a simple comparison `err == ErrBadRequest` can be used to differentiate between the types of error within hystrix-go package.
Now, in order to surface the origin cause of bad request, clients need to use `errors.Wrap` to wrap the original error in ErrBadRequest. It would be possible get the error out from errChan using `err.Cause()`

Reply:
I tend to side more with @epivovar. I'm not sure how using `errors.Wrap` would work to solve this problem, and I think that making BadRequests a concrete type actually makes for a less extensible solution.

I put together this unit tests to try and showcase some of my reservations around making bad request a circuit error. I hope that helps.

```go
//Something like this might work but will still involve type assertions
type circuitErrorWithCause struct {
	CircuitError
	cause error
}

func TestProposal(t *testing.T) {
	badRequest := CircuitError{
		Message: "BadRequest",
	}

	Convey("My issues with the making bad requests a circuit error", t, func() {
		Convey("Wrapped errors will not equal BadRequests", func() {

			wrappedError := errors.Wrap(fmt.Errorf("some bad request"), badRequest.Error())
			So(badRequest == wrappedError, ShouldBeFalse)

		})

		//Here is some code that i could see working
		Convey("Client code is messy", func() {

			//Example handler func
			err := func() error {
				//the client does some work
				err := errors.New("then an error occurs because of bad input")

				//the client then has to check their error to see if it is a bad request
				if err != nil  {
					//then they have to return the bad request so that it doesn't count against the circuit errors
					//this will cover up the original cause
					return badRequest
				}

				return  nil
			}()


			Convey("This is the only way that a simple equality check would work", func() {
				So(err == badRequest, ShouldBeTrue)
			})
		})

		Convey("Adding cause to circuit errors still requires type assertions", func() {
			wrappedError := circuitErrorWithCause{
				CircuitError: badRequest,
				cause: fmt.Errorf("client error"),
			}

			//Example handler func
			err := func() error {
				return wrappedError
			}()

			Convey("Simple Equality check won't work", func() {
				So(err == badRequest, ShouldBeFalse)
			})

			Convey("Would have to assert to the wrapped type so I don't see what we would be gaining", func() {
				assertedType, ok := err.(circuitErrorWithCause)
				So(ok, ShouldBeTrue)
				Convey("You may still need to check the circuit error in the asserted type", func() {
					So(assertedType.CircuitError == badRequest, ShouldBeTrue)
				})
			})
		})
	})
}
```

Overall i think what you want something more akin to the `HystrixBadRequestException` which handlers in java can throw themselves or extend in their own Exception implementations.